presto-hive/src/main/java/com/facebook/presto/hive/util/HiveFileIterator.java

```diff
@@ -25,103 +25,154 @@
 
 import java.io.FileNotFoundException;
 import java.io.IOException;
+import java.util.ArrayDeque;
+import java.util.Collections;
+import java.util.Deque;
+import java.util.Iterator;
 
+import static com.facebook.presto.hadoop.HadoopFileStatus.isDirectory;
 import static com.facebook.presto.hive.HiveErrorCode.HIVE_FILESYSTEM_ERROR;
 import static com.facebook.presto.hive.HiveErrorCode.HIVE_FILE_NOT_FOUND;
 import static java.util.Objects.requireNonNull;
 
 public class HiveFileIterator
         extends AbstractIterator<LocatedFileStatus>
 {
+    public enum NestedDirectoryPolicy
+    {
+        IGNORED,
+        RECURSE,
+        FAIL
+    }
+
+    private final Deque<Path> paths = new ArrayDeque<>();
     private final FileSystem fileSystem;
     private final DirectoryLister directoryLister;
     private final NamenodeStats namenodeStats;
-    private final Path path;
-    private final String partitionName;
-    private final InternalHiveSplitFactory splitFactory;
+    private final NestedDirectoryPolicy nestedDirectoryPolicy;
 
-    private RemoteIterator<LocatedFileStatus> remoteIterator;
+    private Iterator<LocatedFileStatus> remoteIterator = Collections.emptyIterator();
 
     public HiveFileIterator(
             Path path,
             FileSystem fileSystem,
             DirectoryLister directoryLister,
             NamenodeStats namenodeStats,
-            String partitionName,
-            InternalHiveSplitFactory splitFactory)
+            NestedDirectoryPolicy nestedDirectoryPolicy)
     {
-        this.partitionName = requireNonNull(partitionName, "partitionName is null");
-        this.path = requireNonNull(path, "path is null");
+        paths.addLast(requireNonNull(path, "path is null"));
         this.fileSystem = requireNonNull(fileSystem, "fileSystem is null");
         this.directoryLister = requireNonNull(directoryLister, "directoryLister is null");
         this.namenodeStats = requireNonNull(namenodeStats, "namenodeStats is null");
-        this.splitFactory = requireNonNull(splitFactory, "splitFactory is null");
+        this.nestedDirectoryPolicy = requireNonNull(nestedDirectoryPolicy, "nestedDirectoryPolicy is null");
     }
 
     @Override
     protected LocatedFileStatus computeNext()
     {
-        try {
-            if (remoteIterator == null) {
-                remoteIterator = getLocatedFileStatusRemoteIterator(path);
-            }
-
+        while (true) {
             while (remoteIterator.hasNext()) {
                 LocatedFileStatus status = getLocatedFileStatus(remoteIterator);
 
-                // ignore hidden files. Hive ignores files starting with _ and . as well.
+                // Ignore hidden files and directories. Hive ignores files starting with _ and . as well.
```

Comment:
This "and directories" part seems wrong since we aren't always ignoring

Reply:
I believe we are ignoring directories also, because we check this before checking the "directory" flag.